### PR TITLE
Fix MSI upgrade and shortcut issues

### LINKED ISSUE #11995: Pinned shortcut icon disappears on upgrade
When upgrading from v7 RC to RC, my pinned pwsh.exe shortcut on the taskbar (pinned from the [shortcut you create](https://github.com/PowerShell/PowerShell/blob/00b60c44fa91606a7458c3e9a428f8d05c0639d7/assets/Product.wxs#L247)) disappears. This is because Windows Installer icons via WiX's `<Icon>` element get random GUIDs that change from release to release. As a former WiX dev and looong-time setup dev, we recommended people away from those icons and other advertised features because of various problems like this.

You can still use it if you want to customize your Add/Remove Programs (ARP) entry like you do via `ARPPRODUCTICON`, but don't set the `Shortcut@Icon` property to it. Instead, omit the `@Icon` attribute and it will default to the first icon index in the target EXE instead of a cached path to the Windows Installer cache that will change from release to release as you rebuild your MSI.

### LINKED ISSUE #12011: Older/other versions of pwsh are not cleaned up during upgrade to 7.0.0 or manual uninstallation.
I installed the latest and greatest PowerShell 7.0.0 using the x64 MSI. But not all remnants of older/other PowerShell versions are cleaned up.

Both the System environment variables for 7-preview and/or 6 remained intact.
Both context menus for 7-preview and 6 remained intact, though not functioning.

# Steps to reproduce

Install PowerShell 6.2.4 with context menu's enabled.
Install PowerShell 7.0.0 RC3-preview with context menu's enabled.
Install PowerShell 7.0.0 (with context menu's enabled)
(manual) Uninstall PowerShell 7.0.0 RC3

# Expected behavior

No context menu's for PowerShell 6 and/or PowerShell 7-preview.
No system environment variables in $PATH pointing to PowerShell 6 and/or 7-preview.

# Actual behavior

image

# Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.0.0
PSEdition                      Core
GitCommitId                    7.0.0
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by heaths**
Fixes #11995 and fixes #12011 based on changes described in https://github.com/PowerShell/PowerShell/issues/12011#issuecomment-596105112.

### CODE REVIEW

**Comment by iSazonov** [comment]
@heaths Please look CI fail.

**Comment by SteveL-MSFT** [comment]
@heaths take a look at https://dev.azure.com/powershell/PowerShell/_build/results?buildId=53706&view=logs&j=1057fe42-c88b-5deb-88ad-59245c714665&t=1888043b-254d-5761-80ec-47d0d8adc59a, the test detected some file changes, if this is expected, just follow the instructions in the output

**Comment by TravisEz13** [comment]
I updated the wxs for you and filed:  https://github.com/PowerShell/PowerShell/issues/12799

**Comment by heaths** [comment]
@TravisEz13, thanks. I was going to get to this tonight.

Were these actually removed, though? I'm not clear on whether their presence or not is actually a bug. They aren't needed except for `ARPPRODUCTICON` which must reference a row in the `Icon` table but, unless you're advertising your product (almost no one does that anymore - Office stopped over a decade ago), shouldn't be used for shortcuts.

**Comment by iSazonov** [comment]
> Were these actually removed, though?

I wonder too how WiX works - wxs file still contains references to the icons.

**Comment by heaths** [comment]
@iSazonov, what exactly do you mean? As a long-time developer on WiX (I designed and implemented many features in it, most notably package ref-counting that other installers use as well), I could probably explain if you can clarify what you meant.

When the `<Icon>` is used, the binding (during linking, which is light.exe) knows to stream in a file reference; otherwise, when just a normal `<File>` is process the binder puts it in a cabinet that is later left external or streamed into the MSI based on `<Media>` entries (`<MediaTemplate>` provides default behavior and is often easier for most packages).

My main change here was to remove most GUIDs. They aren't needed, except for the `UpgradeCode` property (you normally want that fixed) and shared components without a single key path resource. In fact, just being a shared component - if it always resolves to the same key path - will get the same GUID. This change is mainly about letting WiX create a durable GUID based on the key path at build time, which is far more reliable and easier than people having to maintain it, which wasn't being done here and the main problem with files and registry values not getting cleaned up (they were ref-counted). See https://heaths.dev/setup/2020/05/13/simple-wix-project.html for more explanation and links to many of my old blog posts on what is now the VS Setup team blog (used to be my professional blog - which I used mostly for team purposes - up until a year or so ago).

**Comment by iSazonov** [comment]
> what exactly do you mean?

I mean @TravisEz13 's commit and your question "Were these actually removed, though?".

**Comment by heaths** [comment]
Last I updated the file via `git rebase` they were still there.

**Comment by iSazonov** [comment]
Yes, they are on file system. My concern is that it seems wxs file contains references to the files.

**Comment by heaths** [comment]
They shouldn't anymore. It's possible when I generated my repo wasn't entirely clean. It was a bit behind before I rebased on upstream/master this past weekend since I've been swamped with work stuff.

**Comment by TravisEz13** [comment]
@heaths 
> Were these actually removed, though? I'm not clear on whether their presence or not is actually a bug. 

in CI we generate multiple packages at at time.  There is probably a bug where one of the package generations deletes the icons.  There is code in the packaging to create a staging folder.  We need to make sure all types of packages that add/or remove files from the build use a staging directory.

I filed a bug to investigate.

> I updated the wxs for you and filed: #12799

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
https://github.com/PowerShell/PowerShell/issues/12799 was an existing bug.   We can merge this PR.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by heaths** [comment]
Thanks. If I can find time, there's some more clean-up I'd like to do (you can do almost everything in the .wxs itself, which makes it easier to maintain since it's not spread all over), but wasn't necessary for now. I also recommend passing parameters instead of setting environment variables, as the scope is smaller (though, at least you scope them to the current and child processes).

**Comment by iSazonov** [comment]
@heaths Thanks for great contribution!

**Comment by TravisEz13** [comment]
Thanks for your contribution

File: assets/Product.wxs [hunk 1]
```diff
@@ -5,28 +5,36 @@
...
+    <?if $(var.IsPreview)=True?>
+      <?define UpgradeCode = $(var.UpgradeCodePreview)?>
+    <?else?>
+      <?define UpgradeCode = $(var.UpgradeCodeRelease)?>
+    <?endif?>
   <?else?>
-    <?define UpgradeCode = "$(env.UpgradeCodeX86)" ?>
     <?define ExplorerContextMenuDialogText = "&$(var.ProductName) $(env.SimpleProductVersion) ($(sys.BUILDARCH))"?>
+    <?define UpgradeCodePreview = "1d00683b-0f84-4db8-a64f-2f98ad42fe06"?>
+    <?define UpgradeCodeRelease = "86abcfbd-1ccc-4a88-b8b2-0facfde29094"?>
```
Comment (TravisEz13) [inline assets/Product.wxs:22]: preview upgrade codes
```
39243d76-adaf-42b1-94fb-16ecf83237c8', "Process")	
86abcfbd-1ccc-4a88-b8b2-0facfde29094'
```

Comment (TravisEz13) [reply assets/Product.wxs:22]: release upgrade codes
```
"UpgradeCodeX64", '31ab5147-9a97-4452-8443-d9709f0516e1', "Process")	
"UpgradeCodeX86", '1d00683b-0f84-4db8-a64f-2f98ad42fe06', "Process")
```

Comment (TravisEz13) [reply assets/Product.wxs:22]: Created https://github.com/PowerShell/PowerShell/pull/13036 to fix this

File: tools/packaging/packaging.psm1 [hunk 2]
```diff
@@ -2994,35 +2987,23 @@ function New-MSIPackage
...
-        # The ApplicationProgramsMenuShortcut GUID should be changed when bumping the major version because the installation directory changes
-        [Environment]::SetEnvironmentVariable("ApplicationProgramsMenuShortcut", '6a69de6c-183d-4bf4-a40e-83007d6293bf', "Process")
     }
     else
     {
-        [Environment]::SetEnvironmentVariable("PwshPath", "[$productDirectoryName]preview", "Process")
-        [Environment]::SetEnvironmentVariable("UpgradeCodeX64", '39243d76-adaf-42b1-94fb-16ecf83237c8', "Process")
-        [Environment]::SetEnvironmentVariable("UpgradeCodeX86", '86abcfbd-1ccc-4a88-b8b2-0facfde29094', "Process")
         [Environment]::SetEnvironmentVariable("IconPath", 'assets\Powershell_av_colors.ico', "Process")
-        [Environment]::SetEnvironmentVariable("ApplicationProgramsMenuShortcut", 'ab727c4f-2311-474c-9ade-f2c6fd7f7322', "Process")
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1:3025]: Add comment to Hightlight change

File: tools/packaging/packaging.psm1 [hunk 3]
```diff
@@ -2994,35 +2987,23 @@ function New-MSIPackage
...
-        [Environment]::SetEnvironmentVariable("PwshPath", "[$productDirectoryName]", "Process")
-        [Environment]::SetEnvironmentVariable("UpgradeCodeX64", '31ab5147-9a97-4452-8443-d9709f0516e1', "Process")
-        [Environment]::SetEnvironmentVariable("UpgradeCodeX86", '1d00683b-0f84-4db8-a64f-2f98ad42fe06', "Process")
         [Environment]::SetEnvironmentVariable("IconPath", 'assets\Powershell_black.ico', "Process")
-        # The ApplicationProgramsMenuShortcut GUID should be changed when bumping the major version because the installation directory changes
-        [Environment]::SetEnvironmentVariable("ApplicationProgramsMenuShortcut", '6a69de6c-183d-4bf4-a40e-83007d6293bf', "Process")
     }
     else
     {
-        [Environment]::SetEnvironmentVariable("PwshPath", "[$productDirectoryName]preview", "Process")
```
Comment (TravisEz13) [inline tools/packaging/packaging.psm1:3021]: Add comment to highlight change

**Pull request merged**